pyeapi/api/staticroute.py

```diff
@@ -0,0 +1,302 @@
+                       r'(?: (\d+\.\d+\.\d+\.\d+))?'
+                       r' (\d+)'
+                       r'(?: tag (\d+))?'
+                       r'(?: name (\S+))?', re.M)
+
+# Define a format for the unique route id
+# The four parts in order are ip_dest, next_hop, next_hop_ip, and distance
+ROUTE_ID = "%s--%s--%s--%s"
+
+
+class StaticRoute(EntityCollection):
+    """The StaticRoute class provides a configuration instance
+    for working with static routes
+
+    """
+
+    def __str__(self):
+        return 'StaticRoute'
+
+    def get(self, ip_dest, next_hop, distance, next_hop_ip=None):
+        """Retrieves the ip route information for the route specified
+            by the ip_dest, next_hop, and distance parameters
+
+        Args:
+            ip_dest (string): The ip address of the destination in the
+                form of A.B.C.D/E
+            next_hop (string): The next hop interface or ip address
+            distance (int): Administrative distance for this route
+
+        Returns:
+            dict: An ip route dict object
+
+            If the unique route specified by the ip_dest, next_hop, and
+            distance does not exist, then None is returned.
+        """
+
+        # If distance is None, then set to 1 to match what EOS will
+        # do when distance is not specified.
+        if distance is None:
+            distance = 1
+
+        # Make the unique route_id for the requested route
+        route_id = ROUTE_ID % (ip_dest, next_hop, next_hop_ip, distance)
+
+        # Return the route configuration if found, or return None
+        return self.getall().get(route_id)
+
+    def getall(self):
+        """Return all ip routes configured on the switch as a resource dict
+
+        Returns:
+            dict: A dict of unique ip route names with a nested route
+                dict object. The unique name is built with the ip destination,
+                next hop, and distance values for the route.
+        """
+
+        # Find all the ip routes in the config
+        matches = ROUTES_RE.findall(self.config)
+
+        # Parse the routes and add them to the routes dict
+        routes = dict()
+        for match in matches:
+            # Set the route dict to the returned values, replacing
+            # empty strings with None
+            route = dict()
+            route['ip_dest'] = match[0]
+            route['next_hop'] = match[1]
+            route['next_hop_ip'] = None if match[2] is '' else match[2]
+            route['distance'] = match[3]
+            route['tag'] = None if match[4] is '' else match[4]
+            route['route_name'] = None if match[5] is '' else match[5]
+
+            # Build a unique route_id from the ip_dest, next_hop, and distance
+            route_id = ROUTE_ID % \
+                (route['ip_dest'], route['next_hop'],
+                 route['next_hop_ip'], route['distance'])
+
+            # Update the routes dict
+            routes.update({route_id: route})
```

Comment:
Hi Phil,

This is a great suggestion, and was fairly simple to implement the 
change. I'm still picking up some of the nuances of python, and 
appreciate the tips you have given.

Gary


On 10/15/15 1:46 PM, Philip DiLeo wrote:
>
> In pyeapi/api/staticroute.py 
> <https://github.com/arista-eosplus/pyeapi/pull/45#discussion_r42170898>:
>
> > +            # empty strings with None
> > +            route = dict()
> > +            route['ip_dest'] = match[0]
> > +            route['next_hop'] = match[1]
> > +            route['next_hop_ip'] = None if match[2] is '' else match[2]
> > +            route['distance'] = match[3]
> > +            route['tag'] = None if match[4] is '' else match[4]
> > +            route['route_name'] = None if match[5] is '' else match[5]
> > +
> > +            # Build a unique route_id from the ip_dest, next_hop, and distance
> > +            route_id = ROUTE_ID % \
> > +                (route['ip_dest'], route['next_hop'],
> > +                 route['next_hop_ip'], route['distance'])
> > +
> > +            # Update the routes dict
> > +            routes.update({route_id: route})
>
> Hi Gary,
> I was doing some testing and I'm not sure the hyphen-separated key is 
> the way to go. This produces:
>
> |{'1.1.1.0/24--2.2.2.2--None--1': {'distance': '1', 'ip_dest': 
> '1.1.1.0/24', 'next_hop': '2.2.2.2', 'next_hop_ip': None, 
> 'route_name': None, 'tag': '0'}, '1.1.1.0/24--2.2.2.2--None--3': 
> {'distance': '3', 'ip_dest': '1.1.1.0/24', 'next_hop': '2.2.2.2', 
> 'next_hop_ip': None, 'route_name': None, 'tag': '0'}, 
> '1.1.1.0/24--2.2.2.2--None--5': {'distance': '5', 'ip_dest': 
> '1.1.1.0/24', 'next_hop': '2.2.2.2', 'next_hop_ip': None, 
> 'route_name': None, 'tag': '0'}} |
>
> Which is a little hard to read. For routemaps, I implemented a 
> tuple-based key, eg for staticroute:
>
> |('1.1.1.0/24', '2.2.2.2', None, 5): {'distance': '5', 'ip_dest': 
> '1.1.1.0/24', 'next_hop': '2.2.2.2', 'next_hop_ip': None, 
> 'route_name': None, 'tag': '0'}} |
>
> This is nice because you can use Python's built in get method to do 
> something like:
>
> staticroutes_dict.get( ('1.1.1.0/24', '2.2.2.2', None, 5) ) and it 
> would return the corresponding dictionary.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/arista-eosplus/pyeapi/pull/45/files#r42170898>.
>